service/history/hsm/tree.go

```diff
@@ -0,0 +1,366 @@
+// Returns [ErrStateMachineAlreadyExists] if a child with the given key already exists, [ErrNotRegistered] if the key's
+// type is not found in the node's state machine registry and serialization errors.
+func (n *Node) AddChild(key Key, data any) (*Node, error) {
+	machines, ok := n.persistence.Children[key.Type]
+	if ok {
+		if _, ok = machines.MachinesById[key.ID]; ok {
+			if ok {
+				return nil, fmt.Errorf("%w: %v", ErrStateMachineAlreadyExists, key)
+			}
+		}
+	}
+	def, ok := n.registry.Machine(key.Type)
+	if !ok {
+		return nil, fmt.Errorf("%w: state machine for type: %v", ErrNotRegistered, key.Type)
+	}
+	serialized, err := def.Serialize(data)
+	if err != nil {
+		return nil, err
+	}
+	node := &Node{
+		Key:        key,
+		Parent:     n,
+		definition: def,
+		registry:   n.registry,
+		persistence: &persistencespb.StateMachineNode{
+			Children: make(map[int32]*persistencespb.StateMachineMap),
+			Data:     serialized,
+		},
+		cache: &cachedMachine{
+			dataLoaded: true,
+			data:       data,
+			children:   make(map[Key]*Node),
+		},
+	}
+	n.cache.children[key] = node
+	children, ok := n.persistence.Children[key.Type]
+	if !ok {
+		children = &persistencespb.StateMachineMap{MachinesById: make(map[string]*persistencespb.StateMachineNode)}
+		n.persistence.Children[key.Type] = children
+	}
+	children.MachinesById[key.ID] = node.persistence
+	return node, nil
+}
+
+// MachineData deserializes the persistent state machine's data, casts it to type T, and returns it.
+// Returns an error when deserialization or casting fails.
+func MachineData[T any](n *Node) (T, error) {
+	var t T
+	if n.cache.dataLoaded {
+		if t, ok := n.cache.data.(T); ok {
+			return t, nil
+		}
+		return t, ErrIncompatibleType
+	}
+	a, err := n.definition.Deserialize(n.persistence.Data)
+	if err != nil {
+		return t, err
+	}
+	n.cache.data = a
+	n.cache.dataLoaded = true
+
+	if t, ok := a.(T); ok {
+		return t, nil
+	}
+	return t, ErrIncompatibleType
+}
+
+// TransitionCount returns the transition count for the state machine contained in this node.
+func (n *Node) TransitionCount() int64 {
+	return n.persistence.TransitionCount
+}
+
+// MachineTransition runs the given transitionFn on a machine's data for the given key.
+// It updates the state machine's metadata and marks the entry as dirty in the node's cache.
+// If the transition fails, the changes are rolled back and no state is mutated.
+func MachineTransition[T any](n *Node, transitionFn func(T) (TransitionOutput, error)) (retErr error) {
+	data, err := MachineData[T](n)
+	if err != nil {
+		return err
+	}
+	n.persistence.TransitionCount++
+	// Rollback on error
+	defer func() {
+		if retErr != nil {
+			n.persistence.TransitionCount--
+			// Force reloading data.
+			n.cache.dataLoaded = false
+		}
+	}()
+	output, err := transitionFn(data)
+	if err != nil {
+		return err
+	}
+	serialized, err := n.definition.Serialize(data)
+	if err != nil {
+		return err
+	}
+	n.persistence.Data = serialized
```

Comment:
It doesn't look like it

Reply:
Yes @yycptt. @tdeebswihart what makes you think otherwise?

Reply:
If two calls to `MachineTransition` occur within a single transaction I don't see anywhere we handle rolling back the result of prior calls.

Or am I misunderstanding what a transaction is in this context?

Reply:
Ah I see, we're relying on the caller to call `ClearTransactionState()` in case the transaction fails. We don't _really_ have a transaction abstraction or even documentation on what a transaction is in the context of this library and we really need one.

Reply:
I think I understand now. The impl here always you to retry a failed transition from the last succeed transition. If the entire transaction failed, we will clear the entire in memory state. That's not done by the ClearTransactionState() I think, that one only clears the transition output, now the serialized state. It's (will be) done by implementation of the `Access` method which is not in this PR. Basically `Access` is the one providing the transaction abstraction. Is that correct @bergundy or I am still missing something ? 

One more question, who is responsible for calling `ClearTransactionState()`? Also the impl of `Access` I assume ?

Reply:
Yes, all of this is in the next PR.